Update calendar filter

vargaat opened the pull request:
### What's new?
- Calendar filter screen was completely re-written with the teacher calendar in mind.
- Personal filter is now visible in the filter list and de-selectable, previously we downloaded the personal calendar all the time.
- Calendar filter limit (10/20) is now taken into account for teachers with an info text showing on the filter screen.
- Groups are now filterable.
- Selected calendars are now saved to user defaults instead of CoreData so a breaking DB change won't reset the selections.

### Known issues
- Voiceover not focusing on section header texts or focuses them in an incorrect order based on the iOS version. I couldn't work around this and seems to be an iOS bug.
- Double loading indicators. I had to add the refreshable modifier to the view otherwise the swipe down to dismiss gesture swallowed the pull to refresh gesture and the page wasn't refreshable. Now the page is refreshable but there are two loading indicators: one is our custom one attached to the scrollview and the other one is from the system. The system one quickly disappears and sometimes not even appears so I accepted this workaround. I tried a few different workarounds -even tried to disable the apple private gesture recognizer- but none of them worked.
- Filtering in the teacher calendar is not working correctly at the moment, we have a ticket for the teacher logic so only test the picker UI there.
- If you notice anything strange with filtering compare it to the production version, I tried not to touch the filtering logic here that much since we have a dedicated task for that.

refs: [MBL-17510](https://instructure.atlassian.net/browse/MBL-17510)
affects: Student, Teacher, Parent
release note: Calendar filter screen was updated with new visuals and the capability to filter to personal and group calendars.

test plan: Check calendar filter screen in all three apps.

## Screenshots

<table>
<tr><th>Before</th><th>After</th></tr>
<tr>
<td><img src="https://github.com/instructure/canvas-ios/assets/72396990/cf666754-e16c-4d70-9b28-0bbaba6239c5" maxHeight=500></td>
<td><img src="https://github.com/instructure/canvas-ios/assets/72396990/9c54a407-2fa4-4df7-873d-e25bbe8b2320" maxHeight=500></td>
</tr>
<tr>
<td><img src="https://github.com/instructure/canvas-ios/assets/72396990/bd28e231-0fe5-4295-9d76-e974e5100c7e" maxHeight=500></td>
<td><img src="https://github.com/instructure/canvas-ios/assets/72396990/73890070-7a53-41fa-b6b5-f96e8f087ee4" maxHeight=500></td>
</tr>

</table>

## Checklist

- [ ] Follow-up e2e test ticket created
- [ ] A11y checked
- [ ] Tested on phone
- [ ] Tested on tablet
- [ ] Tested in dark mode
- [ ] Tested in light mode
- [ ] Approve from product

inst-danger commented:
<details><summary>Parent Build QR Code:</summary>
<img src="https://api.qrserver.com/v1/create-qr-code/?data=https%3A%2F%2Fapp.bitrise.io%2Fapp%2F88a3641e29e4745c%2Fbuild%2F5f37d824-aee6-4107-b2c5-23e74d38e5cc%2Fartifact%2Ff2634177ec595c1f%2Fp%2Fc20643d46237199fc33110b4bf3fc398&size=150x150" />
</details>


<!-- ParentBuild -->

inst-danger commented:
<details><summary>Student Build QR Code:</summary>
<img src="https://api.qrserver.com/v1/create-qr-code/?data=https%3A%2F%2Fapp.bitrise.io%2Fapp%2F88a3641e29e4745c%2Fbuild%2Fde0fb54c-1f43-4846-8e34-a303260836ca%2Fartifact%2Fdb05b43c793b2bd5%2Fp%2F38d14450690d3a1965a0885a22738c68&size=150x150" />
</details>


<!-- StudentBuild -->

inst-danger commented:
<details><summary>Teacher Build QR Code:</summary>
<img src="https://api.qrserver.com/v1/create-qr-code/?data=https%3A%2F%2Fapp.bitrise.io%2Fapp%2F88a3641e29e4745c%2Fbuild%2F9899c606-1a58-4683-9871-3660c1ddf0f9%2Fartifact%2F1b8cc909d28833b4%2Fp%2Fb6f2279943d489d83a5366f11616d0ff&size=150x150" />
</details>


<!-- TeacherBuild -->

inst-danger commented:
<!--
  1 failure:  Build failed, ski...
  0 warning: 
  
  4 markdown notices
  DangerID: danger-id-Danger;
-->

<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true">Fails</th>
    </tr>
  </thead>
  <tbody><tr>
      <td>:no_entry_sign:</td>
      <td>Build failed, skipping coverage check</td>
    </tr>
  </tbody>
</table>



#### Release Note: 
Calendar filter screen was updated with new visuals and the capability to filter to personal and group calendars.

#### Affected Apps: Student, Teacher, Parent

[MBL-17510](https://instructure.atlassian.net/browse/MBL-17510)

<details><summary>:x: XCTest failed: CoreTests/CalendarDaysViewControllerTests/testDates</summary>


    XCTAssertEqual failed: ("Optional(0)") is not equal to ("Optional(1)")
    XCTAssertEqual failed: ("Optional(0)") is not equal to ("Optional(2)")
    XCTAssertEqual failed: ("Optional(0)") is not equal to ("Optional(3)")
</details>
<details><summary>:x: XCTest failed: CoreTests/CreateTodoViewControllerTests/testLayout</summary>


    Asynchronous wait failed: Exceeded timeout of 0.5 seconds, with unfulfilled expectations: "make sure create post request is made when done button is pressed".
</details>
<details><summary>:x: XCTest failed: CoreTests/GetPlannablesTests/testMakeRequest</summary>


    XCTAssertEqual failed: ("nil") is not equal to ("Optional(1)")
</details>
<details><summary>:x: XCTest failed: CoreTests/PlannerListViewControllerTests/testErrorState</summary>


    XCTAssertEqual failed: ("true") is not equal to ("false")
</details>
<details><summary>:x: XCTest failed: CoreTests/PlannerListViewControllerTests/testLayout</summary>


    XCTAssertEqual failed: ("false") is not equal to ("true")
    XCTAssertEqual failed: ("nil") is not equal to ("Optional("assignment a")")
    XCTAssertEqual failed: ("nil") is not equal to ("Optional("Assignment Grades")")
    XCTAssertEqual failed: ("nil") is not equal to ("Optional("#ed0000")")
    XCTAssertEqual failed: ("nil") is not equal to ("Optional("May 10, 2024 at 9:26 AM")")
    XCTAssertEqual failed: ("nil") is not equal to ("Optional("1 point")")
    XCTAssertEqual failed: ("nil") is not equal to ("Optional(false)")
    XCTAssertEqual failed: ("nil") is not equal to ("Optional("note")")
    XCTAssertEqual failed: ("nil") is not equal to ("Optional(true)")
    XCTAssertTrue failed
</details>
<details><summary>:x: XCTest failed: CoreTests/PlannerListViewControllerTests/testNavigationToTodo</summary>


    XCTUnwrap failed: expected non-nil value of type "CoreHostingController<ToDoDetailsScreen>"
    Crash: CoreTester (24723) implicit closure #2 in PlannerListViewControllerTests.testNavigationToTodo()
</details>

<p align="right">
  Generated by :no_entry_sign: <a href="https://danger.systems/js">dangerJS</a> against a20572ce8b03cae0053438eba77571af7653285e
</p>

rh12 commented on Core/Core/Planner/CalendarFilter/View/CalendarFilterScreen.swift:
```diff
@@ -0,0 +1,153 @@
+            Button {
+                viewModel.didTapRightNavButton.send()
+            } label: {
+                if let title = viewModel.rightNavButtonTitle {
```
simply `selectAllButtonTitle`, since it is a specific button and we name it like that in Screen property as well?
same for `didTap...`

vargaat replied:
Yeah, I wanted to go in a more generic way but now it seems unnecessary so I renamed these.

rh12 reviewed (commented):
Just some minor comments (I started before I saw it's in draft)

vargaat commented on Core/Core/AppEnvironment/EnvironmentSettings/CDEnvironmentSetting.swift:
```diff
@@ -0,0 +1,52 @@
+    ) -> CDEnvironmentSetting {
+        let predicate = NSCompoundPredicate(andPredicateWithSubpredicates: [
+            NSPredicate(format: "%K == %@", #keyPath(CDEnvironmentSetting.name), item.name),
+        ])
```
Yes, most probably a copy paste remain.

vargaat commented on Core/Core/Extensions/ColorExtensions.swift:
```diff
@@ -27,4 +27,12 @@ public extension Color {
         }
     }
+
+    static var random: Color {
```
Moved to DEBUG builds only, we don't use this on prod yet.

vargaat commented on Core/Core/Planner/CalendarFilter/Model/Interactor/CalendarFilterCountLimit.swift:
```diff
@@ -0,0 +1,53 @@
+
+public enum CalendarFilterCountLimit {
+    case limited(Int)
+    case unlimited
```
Changed it, this makes it easier to work with.

vargaat commented on Core/Core/Assets.xcassets/icons/checkbox.imageset/checkbox.svg:
Fixed by Marci.

rh12 commented:
QA + 1 (iPad @ iOS 17)
Nice job going through this ever growing task, relentlessly (inside joke intended ;-) )

Some findings, all of them are fine to look into in a followup task:
1. Initial deselection issue described in private
2. Sometimes missing calendar colors described in private
3. VoiceOver: Deselected (or Unselected?) state for a calendar could be also useful to read out, not just Selected state
4. VoiceOver: "Selected" may be more useful to read out after the calendar name, not before (but maybe it's just me)
5. VoiceOver: Selecting/Deselecting a calendar doesn't read out which action just happened. User needs to reselect the cell to know which is the current state (or pay attention, maybe people are more used to it than me)
6. VoiceOver: Same for (De)Select All button action

(will approve once I've finished code review)

rh12 commented on Core/Core/Planner/CalendarFilter/ViewModel/CalendarFilterViewModel.swift:
```diff
@@ -0,0 +1,183 @@
+    func showFilterLimitReachedMessage(limit: Int) -> Future<Void, Never> {
+        Future { [weak self] promise in
+            let message = String(localized: "You can only select up to \(limit) calendars.", bundle: .core)
+            self?.showSnack(message)
```
It may be nice to not enqueue these snacks. I mean only show it if it's not already visible.
User may tap repeatedly on calendars to "try to make it work" (at least I did :) ) and getting multiple repeated messages after a some seconds of delay each time feels strange.
I think just checking `visibleSnack` equality might be enough.

vargaat replied:
Good idea. I implemented this in the SnackBarViewModel.

vargaat replied:
I also made the timer reset when a duplicate snack arrives.

vargaat commented:
> 3. VoiceOver: Deselected (or Unselected?) state for a calendar could be also useful to read out, not just Selected state
> 4. VoiceOver: "Selected" may be more useful to read out after the calendar name, not before (but maybe it's just me)
> 5. VoiceOver: Selecting/Deselecting a calendar doesn't read out which action just happened. User needs to reselect the cell to know which is the current state (or pay attention, maybe people are more used to it than me)

I changed the accessibility representation of this cell to a Toggle which greatly improved the voiceover experience.

> 6. VoiceOver: Same for (De)Select All button action

Added a voiceover announcement for the action.

szabinst reviewed (approved):
QA+1
